sophia/src/parser/xml/_handler.rs

```diff
@@ -0,0 +1,767 @@
+    fn check_unique_id(&mut self, id: Term<F::TermData>) -> Result<Term<F::TermData>> {
+        if self.ids.contains(&id) {
+            Err(RdfError::DuplicateId(id.value())).locate_err_with(&self.reader)
+        } else {
+            self.ids.insert(id.clone());
+            Ok(id)
+        }
+    }
+
+    /// Create a new predicate IRI from an XML name (or a RDF metasyntactic element)
+    fn predicate_iri_start(&self, name: &str) -> Result<Term<F::TermData>> {
+        let p = self
+            .scope()
+            .expand_attribute(name)
+            .locate_err_with(&self.reader)?;
+        if p.matches(&rdf::li) {
+            self.parent_scope().new_li().locate_err_with(&self.reader)
+        } else {
+            Ok(p)
+        }
+    }
+
+    /// Retrieve a predicate IRI from an XML name
+    fn predicate_iri_end(&self, name: &str) -> Result<Term<F::TermData>> {
+        let p = self
+            .scope()
+            .expand_attribute(name)
+            .locate_err_with(&self.reader)?;
+        if p.matches(&rdf::li) {
+            self.parent_scope()
+                .current_li()
+                .locate_err_with(&self.reader)
+        } else {
+            Ok(p)
+        }
+    }
+
+    // ---
+
+    /// Create a new `XmlHandler` from the given `quick_xml::Reader`.
+    pub(super) fn new(reader: Reader<B>) -> Self {
+        let factory: Rc<RefCell<F>> = Default::default();
+        Self {
+            reader: XmlReader::from(reader),
+            parents: Vec::new(),
+            scopes: vec![Scope::with_factory_rc(factory.clone())],
+            triples: LinkedList::new(),
+            factory,
+            bnodes: AtomicUsize::new(0),
+            state: vec![ParsingState::Node],
+            ids: HashSet::new(),
+        }
+    }
+
+    /// Create a new `XmlHandler` using the given URL as the top-level `xml:base`.
+    pub(super) fn with_base(reader: Reader<B>, base: Url) -> Self {
+        let mut parser = Self::new(reader);
+        let mut scope = parser.scope_mut();
+        scope.base = Some(base);
+        parser
+    }
+
+    // ---
+
+    pub(super) fn element_start(&mut self, e: &BytesStart) {
+        if let Err(e) = self.enter_scope(e) {
+            self.triples.push_back(Err(e));
+        }
+
+        let step = match self.state.last().unwrap() {
+            ParsingState::Node => self.node_start(e),
+            ParsingState::Predicate => self.predicate_start(e),
+            ParsingState::Resource => self.predicate_start(e),
+            ParsingState::Collection => self.collection_start(e),
+            ParsingState::CollectionItem => self.collection_item_start(e),
+            ParsingState::Literal => unimplemented!("entering element as literal"),
+            ParsingState::Res => match self.reader.decode(e.name()) {
+                Ok(name) => Err(RdfError::UnexpectedEvent {
+                    expected: format!("<{}>", name),
+                    found: "text".to_owned(),
+                })
```

Comment:
@althonos I refactored the lines above for the new error-handling in sophia.

I'm not sure if `expected` and `found` are switched or if it is correct this way. Could you tell me, please?

_Same case l.613_